File: backend/app/models/equipment.py
"""
Equipment models for the welding system backend.
设备管理数据模型
"""
from datetime import datetime, date
from typing import Optional

from sqlalchemy import Column, Integer, String, Text, Float, Boolean, DateTime, Date, ForeignKey, Enum as SQLEnum
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base


class EquipmentType(str, enum.Enum):
    """设备类型"""
    WELDING_MACHINE = "welding_machine"  # 焊接设备
    CUTTING_MACHINE = "cutting_machine"  # 切割设备
    GRINDING_MACHINE = "grinding_machine"  # 打磨设备
    TESTING_EQUIPMENT = "testing_equipment"  # 检测设备
    AUXILIARY_EQUIPMENT = "auxiliary_equipment"  # 辅助设备
    OTHER = "other"  # 其他


class EquipmentStatus(str, enum.Enum):
    """设备状态"""
    OPERATIONAL = "operational"  # 运行中
    IDLE = "idle"  # 空闲
    MAINTENANCE = "maintenance"  # 维护中
    REPAIR = "repair"  # 维修中
    BROKEN = "broken"  # 故障
    RETIRED = "retired"  # 报废


class MaintenanceType(str, enum.Enum):
    """维护类型"""
    ROUTINE = "routine"  # 例行维护
    PREVENTIVE = "preventive"  # 预防性维护
    CORRECTIVE = "corrective"  # 纠正性维护
    EMERGENCY = "emergency"  # 紧急维护


class WorkspaceType(str, enum.Enum):
    """工作区类型"""
    PERSONAL = "personal"  # 个人工作区
    ENTERPRISE = "enterprise"  # 企业工作区


class AccessLevel(str, enum.Enum):
    """数据访问级别"""
    PRIVATE = "private"  # 仅创建者可见
    FACTORY = "factory"  # 同工厂成员可见
    COMPANY = "company"  # 全公司成员可见
    PUBLIC = "public"  # 公开


class Equipment(Base):
    """设备管理模型"""
    
    __tablename__ = "equipment"
    
    # 主键
    id = Column(Integer, primary_key=True, index=True)
    
    # ==================== 数据隔离核心字段 ====================
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True, comment="创建用户ID")
    workspace_type = Column(String(20), nullable=False, default="personal", index=True, comment="工作区类型")
    company_id = Column(Integer, ForeignKey("companies.id", ondelete="CASCADE"), nullable=True, index=True, comment="企业ID")
    factory_id = Column(Integer, ForeignKey("factories.id", ondelete="SET NULL"), nullable=True, index=True, comment="工厂ID")
    
    # 数据访问控制
    is_shared = Column(Boolean, default=False, comment="是否在企业内共享")
    access_level = Column(String(20), default="private", comment="访问级别")
    
    # ==================== 基本信息 ====================
    equipment_code = Column(String(100), nullable=False, index=True, comment="设备编号")
    equipment_name = Column(String(255), nullable=False, comment="设备名称")
    equipment_type = Column(String(50), nullable=False, comment="设备类型")
    category = Column(String(100), comment="设备类别")
    
    # ==================== 制造商信息 ====================
    manufacturer = Column(String(255), comment="制造商")
    brand = Column(String(100), comment="品牌")
    model = Column(String(100), comment="型号")
    serial_number = Column(String(100), unique=True, comment="序列号")
    
    # ==================== 技术参数 ====================
    specifications = Column(Text, comment="技术规格(JSON)")
    rated_power = Column(Float, comment="额定功率(kW)")
    rated_voltage = Column(Float, comment="额定电压(V)")
    rated_current = Column(Float, comment="额定电流(A)")
    max_capacity = Column(Float, comment="最大容量")
    working_range = Column(String(255), comment="工作范围")
    
    # ==================== 采购信息 ====================
    purchase_date = Column(Date, comment="采购日期")
    purchase_price = Column(Float, comment="采购价格")
    currency = Column(String(10), default="CNY", comment="货币")
    supplier = Column(String(255), comment="供应商")
    warranty_period = Column(Integer, comment="保修期(月)")
    warranty_expiry_date = Column(Date, comment="保修到期日期")
    
    # ==================== 位置信息 ====================
    location = Column(String(255), comment="位置")
    workshop = Column(String(100), comment="车间")
    area = Column(String(100), comment="区域")
    
    # ==================== 状态信息 ====================
    status = Column(String(50), default="operational", comment="状态")
    is_active = Column(Boolean, default=True, comment="是否启用")
    is_critical = Column(Boolean, default=False, comment="是否关键设备")
    
    # ==================== 使用信息 ====================
    installation_date = Column(Date, comment="安装日期")
    commissioning_date = Column(Date, comment="投产日期")
    total_operating_hours = Column(Float, default=0, comment="累计运行时长(h)")
    total_maintenance_hours = Column(Float, default=0, comment="累计维护时长(h)")
    last_used_date = Column(Date, comment="最后使用日期")
    usage_count = Column(Integer, default=0, comment="使用次数")
    
    # ==================== 维护信息 ====================
    last_maintenance_date = Column(Date, comment="最后维护日期")
    next_maintenance_date = Column(Date, comment="下次维护日期")
    maintenance_interval_days = Column(Integer, comment="维护间隔(天)")
    maintenance_count = Column(Integer, default=0, comment="维护次数")
    
    # ==================== 检验信息 ====================
    last_inspection_date = Column(Date, comment="最后检验日期")
    next_inspection_date = Column(Date, comment="下次检验日期")
    inspection_interval_days = Column(Integer, comment="检验间隔(天)")
    calibration_date = Column(Date, comment="校准日期")
    calibration_due_date = Column(Date, comment="校准到期日期")
    
    # ==================== 责任人信息 ====================
    responsible_person_id = Column(Integer, ForeignKey("users.id"), comment="责任人ID")
    operator_ids = Column(Text, comment="操作员ID列表(JSON)")
    
    # ==================== 性能指标 ====================
    availability_rate = Column(Float, comment="可用率(%)")
    utilization_rate = Column(Float, comment="利用率(%)")
    failure_rate = Column(Float, comment="故障率")
    mtbf = Column(Float, comment="平均故障间隔时间(h)")
    mttr = Column(Float, comment="平均修复时间(h)")
    
    # ==================== 附加信息 ====================
    description = Column(Text, comment="描述")
    notes = Column(Text, comment="备注")
    manual_url = Column(String(500), comment="使用手册URL")
    images = Column(Text, comment="图片(JSON)")
    documents = Column(Text, comment="相关文档(JSON)")
    tags = Column(Text, comment="标签")
    
    # ==================== 审计字段 ====================
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False, comment="创建人ID")
    updated_by = Column(Integer, ForeignKey("users.id"), nullable=True, comment="更新人ID")
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False, comment="创建时间")
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False, comment="更新时间")
    
    # ==================== 关系 ====================
    # owner = relationship("User", foreign_keys=[user_id], back_populates="equipment")
    # company = relationship("Company", back_populates="equipment")
    # factory = relationship("Factory", back_populates="equipment")
    # maintenance_records = relationship("EquipmentMaintenance", back_populates="equipment", cascade="all, delete-orphan")
    # usage_records = relationship("EquipmentUsage", back_populates="equipment", cascade="all, delete-orphan")
    
    def __repr__(self):
        return f"<Equipment(id={self.id}, code={self.equipment_code}, name={self.equipment_name})>"


class EquipmentMaintenance(Base):
    """设备维护记录模型"""
    
    __tablename__ = "equipment_maintenance_records"
    
    # 主键
    id = Column(Integer, primary_key=True, index=True)
    
    # ==================== 关联信息 ====================
    equipment_id = Column(Integer, ForeignKey("equipment.id", ondelete="CASCADE"), nullable=False, index=True, comment="设备ID")
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True, comment="用户ID")
    company_id = Column(Integer, ForeignKey("companies.id", ondelete="CASCADE"), nullable=True, index=True, comment="企业ID")
    factory_id = Column(Integer, ForeignKey("factories.id", ondelete="SET NULL"), nullable=True, index=True, comment="工厂ID")
    
    # ==================== 维护信息 ====================
    maintenance_code = Column(String(100), comment="维护编号")
    maintenance_type = Column(String(50), nullable=False, comment="维护类型")
    maintenance_category = Column(String(100), comment="维护类别")
    
    # ==================== 时间信息 ====================
    scheduled_date = Column(Date, comment="计划日期")
    start_date = Column(DateTime, nullable=False, comment="开始时间")
    end_date = Column(DateTime, comment="结束时间")
    duration_hours = Column(Float, comment="持续时长(h)")
    
    # ==================== 执行信息 ====================
    technician_id = Column(Integer, ForeignKey("users.id"), comment="技术员ID")
    technician_name = Column(String(100), comment="技术员姓名")
    team_members = Column(Text, comment="团队成员(JSON)")
    
    # ==================== 维护内容 ====================
    maintenance_items = Column(Text, comment="维护项目(JSON)")
    work_description = Column(Text, comment="工作描述")
    parts_replaced = Column(Text, comment="更换部件(JSON)")
    materials_used = Column(Text, comment="使用材料(JSON)")
    
    # ==================== 结果信息 ====================
    status = Column(String(50), default="completed", comment="状态")
    result = Column(String(50), comment="结果")
    issues_found = Column(Text, comment="发现的问题")
    recommendations = Column(Text, comment="建议")
    
    # ==================== 成本信息 ====================
    labor_cost = Column(Float, comment="人工成本")
    parts_cost = Column(Float, comment="配件成本")
    total_cost = Column(Float, comment="总成本")
    currency = Column(String(10), default="CNY", comment="货币")
    
    # ==================== 附加信息 ====================
    notes = Column(Text, comment="备注")
    attachments = Column(Text, comment="附件(JSON)")
    
    # ==================== 审计字段 ====================
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False, comment="创建人ID")
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False, comment="创建时间")
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False, comment="更新时间")
    
    # ==================== 关系 ====================
    # equipment = relationship("Equipment", back_populates="maintenance_records")
    # user = relationship("User", foreign_keys=[user_id])
    # technician = relationship("User", foreign_keys=[technician_id])
    # company = relationship("Company")
    # factory = relationship("Factory")
    
    def __repr__(self):
        return f"<EquipmentMaintenance(id={self.id}, type={self.maintenance_type})>"


class EquipmentUsage(Base):
    """设备使用记录模型"""
    
    __tablename__ = "equipment_usage_records"
    
    # 主键
    id = Column(Integer, primary_key=True, index=True)
    
    # ==================== 关联信息 ====================
    equipment_id = Column(Integer, ForeignKey("equipment.id", ondelete="CASCADE"), nullable=False, index=True, comment="设备ID")
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True, comment="用户ID")
    company_id = Column(Integer, ForeignKey("companies.id", ondelete="CASCADE"), nullable=True, index=True, comment="企业ID")
    factory_id = Column(Integer, ForeignKey("factories.id", ondelete="SET NULL"), nullable=True, index=True, comment="工厂ID")
    
    # 关联业务
    production_task_id = Column(Integer, comment="生产任务ID")
    operator_id = Column(Integer, ForeignKey("users.id"), comment="操作员ID")
    
    # ==================== 使用信息 ====================
    usage_date = Column(Date, nullable=False, comment="使用日期")
    start_time = Column(DateTime, nullable=False, comment="开始时间")
    end_time = Column(DateTime, comment="结束时间")
    duration_hours = Column(Float, comment="使用时长(h)")
    
    # ==================== 工作信息 ====================
    work_type = Column(String(100), comment="工作类型")
    work_description = Column(Text, comment="工作描述")
    output_quantity = Column(Float, comment="产出数量")
    output_unit = Column(String(50), comment="产出单位")
    
    # ==================== 性能信息 ====================
    power_consumption = Column(Float, comment="耗电量(kWh)")
    efficiency = Column(Float, comment="效率(%)")
    quality_rating = Column(Float, comment="质量评分")
    
    # ==================== 问题记录 ====================
    issues_occurred = Column(Boolean, default=False, comment="是否发生问题")
    issue_description = Column(Text, comment="问题描述")
    downtime_hours = Column(Float, comment="停机时长(h)")
    
    # ==================== 附加信息 ====================
    notes = Column(Text, comment="备注")
    
    # ==================== 审计字段 ====================
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False, comment="创建人ID")
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False, comment="创建时间")
    
    # ==================== 关系 ====================
    # equipment = relationship("Equipment", back_populates="usage_records")
    # operator = relationship("User", foreign_keys=[operator_id])
    
    def __repr__(self):
        return f"<EquipmentUsage(id={self.id}, date={self.usage_date})>"

File: frontend/src/services/equipment.ts
import { apiService } from './api'
import { ApiResponse, PaginatedResponse } from '@/types'

// 设备相关类型定义
export interface Equipment {
  id: string
  user_id: string
  company_id?: string
  factory_id?: string
  equipment_code: string
  equipment_name: string
  equipment_type: EquipmentType
  category?: string
  manufacturer?: string
  brand?: string
  model?: string
  serial_number?: string
  specifications?: string
  rated_power?: number
  rated_voltage?: number
  rated_current?: number
  max_capacity?: number
  working_range?: string
  purchase_date?: string
  purchase_price?: number
  currency: string
  supplier?: string
  warranty_period?: number
  warranty_expiry_date?: string
  location?: string
  workshop?: string
  area?: string
  status: EquipmentStatus
  is_active: boolean
  is_critical: boolean
  installation_date?: string
  commissioning_date?: string
  total_operating_hours?: number
  total_maintenance_hours?: number
  last_used_date?: string
  usage_count?: number
  last_maintenance_date?: string
  next_maintenance_date?: string
  maintenance_interval_days?: number
  maintenance_count?: number
  last_inspection_date?: string
  next_inspection_date?: string
  inspection_interval_days?: number
  calibration_date?: string
  calibration_due_date?: string
  responsible_person_id?: number
  operator_ids?: string
  availability_rate?: number
  utilization_rate?: number
  failure_rate?: number
  mtbf?: number
  mttr?: number
  description?: string
  notes?: string
  manual_url?: string
  images?: string
  documents?: string
  tags?: string
  access_level: string
  created_at: string
  updated_at: string
}

export type EquipmentType =
  | 'welding_machine'      // 焊接设备
  | 'cutting_machine'      // 切割设备
  | 'grinding_machine'     // 打磨设备
  | 'testing_equipment'    // 检测设备
  | 'auxiliary_equipment'  // 辅助设备
  | 'other'               // 其他

export type EquipmentStatus =
  | 'operational'  // 运行中
  | 'idle'         // 空闲
  | 'maintenance'  // 维护中
  | 'repair'       // 维修中
  | 'broken'       // 故障
  | 'retired'      // 报废

export type AccessLevel =
  | 'private'  // 仅创建者可见
  | 'factory'  // 同工厂成员可见
  | 'company'  // 全公司成员可见
  | 'public'   // 公开

// 设备创建数据
export interface CreateEquipmentData {
  equipment_code: string
  equipment_name: string
  equipment_type: EquipmentType
  category?: string
  manufacturer?: string
  brand?: string
  model?: string
  serial_number?: string
  specifications?: string
  rated_power?: number
  rated_voltage?: number
  rated_current?: number
  max_capacity?: number
  working_range?: string
  purchase_date?: string
  purchase_price?: number
  currency?: string
  supplier?: string
  warranty_period?: number
  warranty_expiry_date?: string
  location?: string
  workshop?: string
  area?: string
  status?: EquipmentStatus
  is_active?: boolean
  is_critical?: boolean
  installation_date?: string
  commissioning_date?: string
  maintenance_interval_days?: number
  inspection_interval_days?: number
  responsible_person_id?: number
  description?: string
  notes?: string
  manual_url?: string
  images?: string
  documents?: string
  tags?: string
  access_level?: AccessLevel
}

// 设备更新数据
export interface UpdateEquipmentData {
  equipment_name?: string
  equipment_type?: EquipmentType
  category?: string
  manufacturer?: string
  brand?: string
  model?: string
  specifications?: string
  rated_power?: number
  rated_voltage?: number
  rated_current?: number
  max_capacity?: number
  working_range?: string
  purchase_price?: number
  supplier?: string
  location?: string
  workshop?: string
  area?: string
  status?: EquipmentStatus
  is_active?: boolean
  is_critical?: boolean
  description?: string
  notes?: string
  manual_url?: string
  images?: string
  documents?: string
  tags?: string
  access_level?: AccessLevel
}

// 设备状态更新数据
export interface StatusUpdateData {
  status: EquipmentStatus
  notes?: string
}

// 设备查询参数
export interface EquipmentListParams {
  skip?: number
  limit?: number
  search?: string
  equipment_type?: EquipmentType
  status?: EquipmentStatus
  factory_id?: number
  workspace_type?: 'personal' | 'company'
}

// 设备统计信息
export interface EquipmentStatistics {
  total_equipment: number
  operational: number
  idle: number
  maintenance: number
  repair: number
  broken: number
  upcoming_maintenance: number
  overdue_inspection: number
  status_counts: Record<string, number>
  type_counts: Record<string, number>
}

// 维护提醒
export interface MaintenanceAlert {
  id: string
  equipment_code: string
  equipment_name: string
  equipment_type: EquipmentType
  next_maintenance_date: string
  days_until_maintenance: number
  urgency: 'urgent' | 'normal' | 'low'
  location?: string
  status: EquipmentStatus
}

export interface MaintenanceRecordItem {
  id: number
  equipment_id: number
  equipment_code: string
  equipment_name: string
  maintenance_code?: string
  maintenance_type: string
  start_date?: string
  end_date?: string
  duration_hours?: number
  technician_name?: string
  work_description?: string
  result?: string
  status?: string
  notes?: string
  created_at?: string
}

export interface UsageRecordItem {
  id: number
  equipment_id: number
  equipment_code: string
  equipment_name: string
  usage_date?: string
  start_time?: string
  end_time?: string
  duration_hours?: number
  operator_id?: number
  work_type?: string
  work_description?: string
  output_quantity?: number
  output_unit?: string
  issues_occurred?: boolean
  issue_description?: string
  notes?: string
  created_at?: string
}

export interface CreateMaintenanceRecordData {
  equipment_id: number
  maintenance_type: string
  start_date: string
  end_date?: string
  duration_hours?: number
  technician_name?: string
  work_description?: string
  result?: string
  notes?: string
}

export interface CreateUsageRecordData {
  equipment_id: number
  usage_date: string
  start_time?: string
  end_time?: string
  duration_hours?: number
  work_type?: string
  work_description?: string
  output_quantity?: number
  output_unit?: string
  issues_occurred?: boolean
  issue_description?: string
  notes?: string
}

// 设备服务类
class EquipmentService {
  // ==================== 设备基础管理 ====================

  /**
   * 获取设备列表
   */
  async getEquipmentList(params?: EquipmentListParams): Promise<PaginatedResponse<Equipment>> {
    const queryParams = new URLSearchParams()

    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
    if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
    if (params?.search) queryParams.append('search', params.search)
    if (params?.equipment_type) queryParams.append('equipment_type', params.equipment_type)
    if (params?.status) queryParams.append('status', params.status)
    if (params?.factory_id) queryParams.append('factory_id', params.factory_id.toString())
    if (params?.workspace_type) queryParams.append('workspace_type', params.workspace_type)

    const url = `/equipment/${queryParams.toString() ? `?${queryParams.toString()}` : ''}`
    const response = await apiService.get<PaginatedResponse<Equipment>>(url)

    return response.data
  }

  /**
   * 获取设备详情
   */
  async getEquipmentDetail(equipmentId: string): Promise<ApiResponse<Equipment>> {
    const response = await apiService.get<Equipment>(`/equipment/${equipmentId}`)
    return response.data
  }

  /**
   * 创建新设备
   */
  async createEquipment(data: CreateEquipmentData): Promise<ApiResponse<Equipment>> {
    // 获取当前工作区类型
    const currentWorkspace = this.getCurrentWorkspace()
    const workspaceType = currentWorkspace?.type === 'enterprise' ? 'company' : 'personal'

    // 添加工作区类型参数到URL (注意URL末尾的斜杠,避免307重定向)
    const response = await apiService.post<Equipment>(
      `/equipment/?workspace_type=${workspaceType}`,
      data
    )
    return response.data
  }

  /**
   * 获取当前工作区（从本地存储）
   */
  private getCurrentWorkspace() {
    try {
      const workspaceData = localStorage.getItem('current_workspace')
      return workspaceData ? JSON.parse(workspaceData) : null
    } catch (error) {
      console.error('获取当前工作区失败:', error)
      return null
    }
  }

  /**
   * 更新设备信息
   */
  async updateEquipment(equipmentId: string, data: UpdateEquipmentData): Promise<ApiResponse<Equipment>> {
    // 获取当前工作区类型
    const currentWorkspace = this.getCurrentWorkspace()
    const workspaceType = currentWorkspace?.type === 'enterprise' ? 'company' : 'personal'

    const response = await apiService.put<Equipment>(
      `/equipment/${equipmentId}?workspace_type=${workspaceType}`,
      data
    )
    return response.data
  }

  /**
   * 删除设备
   */
  async deleteEquipment(equipmentId: string): Promise<ApiResponse<void>> {
    // 获取当前工作区类型
    const currentWorkspace = this.getCurrentWorkspace()
    const workspaceType = currentWorkspace?.type === 'enterprise' ? 'company' : 'personal'

    const response = await apiService.delete<void>(
      `/equipment/${equipmentId}?workspace_type=${workspaceType}`
    )
    return response.data
  }

  // ==================== 设备状态管理 ====================

  /**
   * 更新设备状态
   */
  async updateEquipmentStatus(equipmentId: string, data: StatusUpdateData): Promise<ApiResponse<Equipment>> {
    const response = await apiService.put<Equipment>(`/equipment/${equipmentId}/status`, data)
    return response.data
  }

  // ==================== 设备统计和提醒 ====================

  /**
   * 获取设备统计信息
   */
  async getEquipmentStatistics(): Promise<ApiResponse<EquipmentStatistics>> {
    // 获取当前工作区类型
    const currentWorkspace = this.getCurrentWorkspace()
    const workspaceType = currentWorkspace?.type === 'enterprise' ? 'company' : 'personal'

    const response = await apiService.get<EquipmentStatistics>(
      `/equipment/statistics/overview?workspace_type=${workspaceType}`
    )
    return response.data
  }

  /**
   * 获取维护提醒
   */
  async getMaintenanceAlerts(days: number = 30): Promise<ApiResponse<{ items: MaintenanceAlert[], total: number }>> {
    const response = await apiService.get<{ items: MaintenanceAlert[], total: number }>(`/equipment/maintenance/alerts?days=${days}`)
    return response.data
  }

  async getMaintenanceRecords(params?: {
    skip?: number
    limit?: number
    equipment_id?: number
  }): Promise<ApiResponse<{ items: MaintenanceRecordItem[], total: number }>> {
    const queryParams = new URLSearchParams()
    const currentWorkspace = this.getCurrentWorkspace()
    queryParams.append('workspace_type', currentWorkspace?.type === 'enterprise' ? 'company' : 'personal')
    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
    if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
    if (params?.equipment_id !== undefined) queryParams.append('equipment_id', params.equipment_id.toString())
    const response = await apiService.get<{ items: MaintenanceRecordItem[], total: number }>(
      `/equipment/maintenance-records?${queryParams.toString()}`
    )
    return response.data
  }

  async createMaintenanceRecord(
    data: CreateMaintenanceRecordData
  ): Promise<ApiResponse<MaintenanceRecordItem>> {
    const currentWorkspace = this.getCurrentWorkspace()
    const workspaceType = currentWorkspace?.type === 'enterprise' ? 'company' : 'personal'
    const response = await apiService.post<MaintenanceRecordItem>(
      `/equipment/maintenance-records?workspace_type=${workspaceType}`,
      data
    )
    return response.data
  }

  async getUsageRecords(params?: {
    skip?: number
    limit?: number
    equipment_id?: number
  }): Promise<ApiResponse<{ items: UsageRecordItem[], total: number }>> {
    const queryParams = new URLSearchParams()
    const currentWorkspace = this.getCurrentWorkspace()
    queryParams.append('workspace_type', currentWorkspace?.type === 'enterprise' ? 'company' : 'personal')
    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
    if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
    if (params?.equipment_id !== undefined) queryParams.append('equipment_id', params.equipment_id.toString())
    const response = await apiService.get<{ items: UsageRecordItem[], total: number }>(
      `/equipment/usage-records?${queryParams.toString()}`
    )
    return response.data
  }

  async createUsageRecord(data: CreateUsageRecordData): Promise<ApiResponse<UsageRecordItem>> {
    const currentWorkspace = this.getCurrentWorkspace()
    const workspaceType = currentWorkspace?.type === 'enterprise' ? 'company' : 'personal'
    const response = await apiService.post<UsageRecordItem>(
      `/equipment/usage-records?workspace_type=${workspaceType}`,
      data
    )
    return response.data
  }

  // ==================== 工具方法 ====================

  /**
   * 获取设备类型选项
   */
  getEquipmentTypeOptions(): Array<{ label: string, value: EquipmentType }> {
    return [
      { label: '焊接设备', value: 'welding_machine' },
      { label: '切割设备', value: 'cutting_machine' },
      { label: '打磨设备', value: 'grinding_machine' },
      { label: '检测设备', value: 'testing_equipment' },
      { label: '辅助设备', value: 'auxiliary_equipment' },
      { label: '其他', value: 'other' }
    ]
  }

  /**
   * 获取设备状态选项
   */
  getEquipmentStatusOptions(): Array<{ label: string, value: EquipmentStatus, color: string }> {
    return [
      { label: '运行中', value: 'operational', color: 'green' },
      { label: '空闲', value: 'idle', color: 'blue' },
      { label: '维护中', value: 'maintenance', color: 'orange' },
      { label: '维修中', value: 'repair', color: 'red' },
      { label: '故障', value: 'broken', color: 'red' },
      { label: '报废', value: 'retired', color: 'gray' }
    ]
  }

  /**
   * 获取访问级别选项
   */
  getAccessLevelOptions(): Array<{ label: string, value: AccessLevel, description: string }> {
    return [
      { label: '私有', value: 'private', description: '仅创建者可见' },
      { label: '工厂级', value: 'factory', description: '同工厂成员可见' },
      { label: '公司级', value: 'company', description: '全公司成员可见' },
      { label: '公开', value: 'public', description: '所有企业成员可见' }
    ]
  }

  /**
   * 格式化设备状态显示
   */
  formatEquipmentStatus(status: EquipmentStatus): { text: string, color: string } {
    const statusMap = {
      operational: { text: '运行中', color: 'green' },
      idle: { text: '空闲', color: 'blue' },
      maintenance: { text: '维护中', color: 'orange' },
      repair: { text: '维修中', color: 'red' },
      broken: { text: '故障', color: 'red' },
      retired: { text: '报废', color: 'gray' }
    }
    return statusMap[status] || { text: status, color: 'default' }
  }

  /**
   * 格式化设备类型显示
   */
  formatEquipmentType(type: EquipmentType): string {
    const typeMap = {
      welding_machine: '焊接设备',
      cutting_machine: '切割设备',
      grinding_machine: '打磨设备',
      testing_equipment: '检测设备',
      auxiliary_equipment: '辅助设备',
      other: '其他'
    }
    return typeMap[type] || type
  }

  /**
   * 生成设备编号
   */
  generateEquipmentCode(type: EquipmentType, companyId?: string): string {
    const now = new Date()
    const year = now.getFullYear().toString().slice(-2)
    const month = (now.getMonth() + 1).toString().padStart(2, '0')
    const day = now.getDate().toString().padStart(2, '0')

    const typePrefix = {
      welding_machine: 'WM',
      cutting_machine: 'CM',
      grinding_machine: 'GM',
      testing_equipment: 'TE',
      auxiliary_equipment: 'AE',
      other: 'OT'
    }

    const prefix = typePrefix[type] || 'EQ'
    const companySuffix = companyId ? `-${companyId.slice(-4)}` : ''

    return `${prefix}-${year}${month}${day}${companySuffix}`
  }

  /**
   * 验证设备数据
   */
  validateEquipmentData(data: CreateEquipmentData): { isValid: boolean, errors: string[] } {
    const errors: string[] = []

    // 必填字段验证
    if (!data.equipment_code?.trim()) {
      errors.push('设备编号不能为空')
    }
    if (!data.equipment_name?.trim()) {
      errors.push('设备名称不能为空')
    }
    if (!data.equipment_type) {
      errors.push('设备类型不能为空')
    }

    // 格式验证
    if (data.purchase_price && (data.purchase_price < 0 || isNaN(data.purchase_price))) {
      errors.push('采购价格必须是正数')
    }
    if (data.rated_power && (data.rated_power < 0 || isNaN(data.rated_power))) {
      errors.push('额定功率必须是正数')
    }
    if (data.rated_voltage && (data.rated_voltage < 0 || isNaN(data.rated_voltage))) {
      errors.push('额定电压必须是正数')
    }
    if (data.rated_current && (data.rated_current < 0 || isNaN(data.rated_current))) {
      errors.push('额定电流必须是正数')
    }

    // 日期验证
    if (data.purchase_date && !this.isValidDate(data.purchase_date)) {
      errors.push('采购日期格式不正确')
    }
    if (data.warranty_expiry_date && !this.isValidDate(data.warranty_expiry_date)) {
      errors.push('保修到期日期格式不正确')
    }
    if (data.installation_date && !this.isValidDate(data.installation_date)) {
      errors.push('安装日期格式不正确')
    }

    return {
      isValid: errors.length === 0,
      errors
    }
  }

  /**
   * 验证日期格式
   */
  private isValidDate(dateString: string): boolean {
    const date = new Date(dateString)
    return date instanceof Date && !isNaN(date.getTime())
  }
}

// 导出单例
export const equipmentService = new EquipmentService()
export default equipmentService